java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java

```diff
@@ -0,0 +1,154 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.grid.web.servlet;
+
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.net.MediaType;
+
+import org.openqa.grid.common.exception.GridException;
+import org.openqa.grid.internal.GridRegistry;
+import org.openqa.grid.internal.RemoteProxy;
+import org.openqa.selenium.json.Json;
+import org.openqa.selenium.json.JsonException;
+import org.openqa.selenium.json.JsonOutput;
+import org.openqa.selenium.remote.http.HttpMethod;
+import org.openqa.selenium.remote.http.HttpRequest;
+import org.openqa.selenium.remote.http.HttpResponse;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.Writer;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * API to query all the sessions that are currently running in the hub.
+ */
+public class NodeSessionsServlet extends RegistryBasedServlet {
+
+  private final Json json = new Json();
+
+  public NodeSessionsServlet() {
+    this(null);
+  }
+
+  public NodeSessionsServlet(GridRegistry registry) {
+    super(registry);
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse rsp) throws IOException {
+    process(rsp);
+  }
+
+  @Override
+  protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+    if (req.getInputStream() != null) {
+      process(resp);
+    } else {
+      process(resp);
+    }
+  }
+
+  protected void process(HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    response.setCharacterEncoding("UTF-8");
+    response.setStatus(200);
+    Map<String, Object> proxies = new TreeMap<>();
+    try (Writer writer = response.getWriter(); JsonOutput out = json.newOutput(writer)) {
+      proxies.put("success", true);
+      proxies.put("proxies", extractSessionsFromAllProxies());
+      out.write(proxies);
+    }
+  }
+
+  private List<Map<String, Object>> extractSessionsFromAllProxies() {
+    List<Map<String, Object>> results = new LinkedList<>();
+    List<RemoteProxy> proxies = getRegistry().getAllProxies().getBusyProxies();
+    for (RemoteProxy proxy : proxies) {
+      Map<String, Object> sessionsInProxy = new TreeMap<>(extractSessionInfo(proxy));
+      if (sessionsInProxy.isEmpty()) {
+        continue;
+      }
+      Map<String, Object> res = new TreeMap<>();
+      res.put("proxyId", proxy.getId());
```

Comment:
This is already going to be returned as a list of proxies. Having the word "`proxy`" seems redundant.

Reply:
The current sessions payload, has no information about the proxy. That is why added both the proxyId and proxyRemoteHost so that the user gets to know as to for which proxy does the sessions info belong to.